File: SPRING/Hospital_Management_system_angular/src/app/services/billing/billing-invoice.service.ts
import { Injectable } from '@angular/core';
import { environment } from '../../../environments/environment';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';

@Injectable({ providedIn: 'root' })
export class BillingInvoiceService {
  private api = environment.apiUrl + 'billing-invoices';

  constructor(private http: HttpClient) {}

  createInvoice(data: any): Observable<any> {
    return this.http.post<any>(this.api, data);
  }

  getById(id: number): Observable<any> {
    return this.http.get<any>(`${this.api}/${id}`);
  }

  getByInvoiceNumber(invoiceNumber: string): Observable<any> {
    return this.http.get<any>(`${this.api}/number/${invoiceNumber}`);
  }

  getAll(): Observable<any[]> {
    return this.http.get<any[]>(this.api);
  }

  getByPatientId(patientId: number): Observable<any[]> {
    return this.http.get<any[]>(`${this.api}/patient/${patientId}`);
  }

  getByAdmittedPatientId(admittedPatientId: number): Observable<any[]> {
    return this.http.get<any[]>(`${this.api}/admission/${admittedPatientId}`);
  }

  searchInvoices(search: string): Observable<any[]> {
    return this.http.get<any[]>(`${this.api}/search`, { params: { search } });
  }

  updateInvoice(id: number, data: any): Observable<any> {
    return this.http.put<any>(`${this.api}/${id}`, data);
  }

  addItem(invoiceId: number, item: any): Observable<any> {
    return this.http.post<any>(`${this.api}/${invoiceId}/items`, item);
  }

  removeItem(invoiceId: number, itemId: number): Observable<any> {
    return this.http.delete<any>(`${this.api}/${invoiceId}/items/${itemId}`);
  }

  finalizeInvoice(id: number, finalizedBy?: string): Observable<any> {
    let params = new HttpParams();
    if (finalizedBy) params = params.set('finalizedBy', finalizedBy);
    return this.http.put<any>(`${this.api}/${id}/finalize`, {}, { params });
  }

  cancelInvoice(id: number): Observable<any> {
    return this.http.put<any>(`${this.api}/${id}/cancel`, {});
  }

  syncFromModules(admittedPatientId: number): Observable<any> {
    return this.http.post<any>(`${this.api}/sync/${admittedPatientId}`, {});
  }

  processPayment(data: any): Observable<any> {
    return this.http.post<any>(`${this.api}/payments`, data);
  }

  getPaymentsByInvoice(invoiceId: number): Observable<any[]> {
    return this.http.get<any[]>(`${this.api}/${invoiceId}/payments`);
  }

  getChargeCategories(): Observable<any[]> {
    return this.http.get<any[]>(`${this.api}/categories`);
  }

  getDashboardSummary(): Observable<any> {
    return this.http.get<any>(`${this.api}/dashboard-summary`);
  }
}

File: SPRING_BOOT/Modern_Hospital_Management_System/src/main/java/emranhss/com/Modern_Hospital_Management_System/service/BillingInvoiceService.java
package emranhss.com.Modern_Hospital_Management_System.service;

import emranhss.com.Modern_Hospital_Management_System.dto.request.BillingInvoiceRequest;
import emranhss.com.Modern_Hospital_Management_System.dto.request.BillingPaymentRequest;
import emranhss.com.Modern_Hospital_Management_System.dto.response.BillingDashboardSummaryResponse;
import emranhss.com.Modern_Hospital_Management_System.dto.response.BillingInvoiceResponse;
import emranhss.com.Modern_Hospital_Management_System.dto.response.BillingPaymentResponse;

import java.util.List;

public interface BillingInvoiceService {

    BillingInvoiceResponse createInvoice(BillingInvoiceRequest request);

    BillingInvoiceResponse getById(Long id);

    BillingInvoiceResponse getByInvoiceNumber(String invoiceNumber);

    List<BillingInvoiceResponse> getAllInvoices();

    List<BillingInvoiceResponse> getByPatientId(Long patientId);

    List<BillingInvoiceResponse> getByAdmittedPatientId(Long admittedPatientId);

    List<BillingInvoiceResponse> searchInvoices(String search);

    BillingInvoiceResponse updateInvoice(Long id, BillingInvoiceRequest request);

    BillingInvoiceResponse addItem(Long invoiceId, BillingInvoiceRequest.BillingInvoiceItemRequest item);

    BillingInvoiceResponse removeItem(Long invoiceId, Long itemId);

    BillingInvoiceResponse finalizeInvoice(Long id, String finalizedBy);

    BillingInvoiceResponse cancelInvoice(Long id);

    BillingInvoiceResponse syncFromModules(Long admittedPatientId);

    BillingPaymentResponse processPayment(BillingPaymentRequest request);

    List<BillingPaymentResponse> getPaymentsByInvoiceId(Long invoiceId);

    BillingDashboardSummaryResponse getDashboardSummary();

    byte[] generatePdf(Long id);
}
